farm/infer.py

```diff
@@ -110,27 +119,48 @@ def load(cls, load_dir, batch_size=4, gpu=False, embedder_only=False, return_cla
         name = os.path.basename(load_dir)
         return cls(model, processor, batch_size=batch_size, gpu=gpu, name=name, return_class_probs=return_class_probs)
 
-    def run_inference(self, dicts):
-        """
-        Runs down-stream inference using the prediction head.
+    def inference_from_file(self, file):
+        dicts = self.processor.file_to_dicts(file)
 
-        :param dicts: Samples to run inference on provided as a list of dicts. One dict per sample.
-        :type dicst: [dict]
-        :return: dict of predictions
+        dict_batches_to_process = int(len(dicts) / self.multiprocessing_chunk_size)
+        num_cpus = min(mp.cpu_count(), dict_batches_to_process) or 1
 
-        """
-        if self.prediction_type == "embedder":
-            raise TypeError(
-                "You have called run_inference for a model without any prediction head! "
-                "If you want to: "
-                "a) ... extract vectors from the language model: call `Inferencer.extract_vectors(...)`"
-                f"b) ... run inference on a downstream task: make sure your model path {self.name} contains a saved prediction head"
+        with ExitStack() as stack:
+            p = stack.enter_context(mp.Pool(processes=num_cpus))
+
+            logger.info(
+                f"Got ya {num_cpus} parallel workers to do inference on {len(dicts)}dicts (chunksize = {self.multiprocessing_chunk_size})..."
             )
-        dataset, tensor_names = self.processor.dataset_from_dicts(dicts, from_inference=True)
+            log_ascii_workers(num_cpus, logger)
+
+            results = p.imap(
+                partial(self._multiproc_dict_to_samples, processor=self.processor),
+                dicts,
+                chunksize=1,
+            )
+
+            samples = []
+            datasets = []
+            for dataset, tensor_names, sample in tqdm(results, total=dict_batches_to_process):
+                datasets.append(dataset)
+                samples.extend(sample)
+
+            concat_datasets = ConcatDataset(datasets)
```

Comment:
For large files it might blow up memory because we keep all "datasets" and all "samples" in memory. It might be an alternative to also call "_run_inference" on the chunks only and concatenate the resulting "preds"